main.go

```diff
@@ -0,0 +1,185 @@
+		failf("* Failed to restart SSH Agent, %s", err)
+	}
+
+	if err := checkPassphrase(cfg.SSHKeySavePath); err != nil {
+		fmt.Println()
+		failf("Error, %s", err)
+	}
+
+	fmt.Println()
+	log.Printf("# Success")
+	log.Printf("The SSH key was saved to *%s*", cfg.SSHKeySavePath)
+	log.Printf("and was successfully added to ssh-agent.")
+}
+
+func ensureSavePath(savePath string) error {
+	dirpath := filepath.Dir(savePath)
+	return os.MkdirAll(dirpath, 0600)
+}
+
+func writeSSHKey(savePath string, SSHRsaPrivateKey string) error {
+	f, err := os.Create(savePath)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		cerr := f.Close()
+		if cerr != nil {
+			log.Errorf("Failed to close the file, err: %s", cerr)
+		}
+	}()
+
+	_, err = f.WriteString(SSHRsaPrivateKey)
+	return err
+}
+
+func restartAgent(removeOtherIdentities bool) error {
+	var shouldStartNewAgent bool
+	cmd := command.New("ssh-add", "-l")
+	cmd.SetStderr(os.Stderr)
+	log.Printf("-> %s", cmd.PrintableCommandArgs())
+
+	returnValue, err := cmd.RunAndReturnExitCode()
+	if err != nil {
+		log.Debugf("Exit code: %s", err)
+	}
+
+	if returnValue == 2 {
+		log.Printf(" (i) ssh_agent_check_result: %d", returnValue)
+		log.Printf(" (i) ssh-agent not started")
+		shouldStartNewAgent = true
+	} else {
+		// ssh-agent loaded and accessible
+		log.Printf(" (i) ssh_agent_check_result: %d", returnValue)
+		fmt.Printf(" (i) running / accessible ssh-agent detected")
+		if removeOtherIdentities {
+			// remove all keys from the current agent
+			cmdRemove := command.New("ssh-add", "-D")
+			cmdRemove.SetStdout(os.Stdout)
+			cmdRemove.SetStderr(os.Stderr)
+
+			fmt.Println()
+			fmt.Println()
+			log.Printf("-> %s", cmdRemove.PrintableCommandArgs())
+
+			if err := cmdRemove.Run(); err != nil {
+				return err
+			}
+
+			// try to kill the agent
+			cmdKill := command.New("ssh-agent", "-k")
+			cmdKill.SetStdout(os.Stdout)
+			cmdKill.SetStderr(os.Stderr)
+
+			fmt.Println()
+			log.Printf("-> %s", cmdKill.PrintableCommandArgs())
+
+			returnValue, err := cmdKill.RunAndReturnExitCode()
+			if err != nil {
+				log.Debugf("Exit code: %s", err)
+			}
+
+			if returnValue == 0 {
+				shouldStartNewAgent = true
+			}
+		}
+	}
+
+	if shouldStartNewAgent {
+		fmt.Printf(" (i) starting a new ssh-agent and exporting connection information with envman")
+		cmd := command.New("ssh-agent")
+
+		fmt.Println()
+		log.Printf("-> %s", cmd.PrintableCommandArgs())
+
+		returnValue, err := cmd.RunAndReturnTrimmedOutput()
+		if err != nil {
+			log.Debugf("Exit code: %s", err)
+		}
+
+		fmt.Printf(" (i) Expose SSH_AUTH_SOCK for the new ssh-agent, with envman")
+
+		returnValue = strings.TrimLeft(returnValue, "SSH_AUTH_SOCK=")
+		returnValue = strings.Split(returnValue, ";")[0]
+
+		if err = os.Unsetenv("SSH_AUTH_SOCK"); err != nil {
+			return fmt.Errorf("Failed to set SSH_AUTH_SOCK env")
+		}
+
+		return tools.ExportEnvironmentWithEnvman("SSH_AUTH_SOCK", returnValue)
+	}
+	return nil
```

Comment:
what if shouldStartNewAgent = false? should not we export anything?

Reply:
Let's discuss this.